master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -653,11 +678,39 @@ private[celeborn] class Master(
       override def run(): Unit = {
         statusSystem.handleAppLost(appId, requestId)
         logInfo(s"Removed application $appId")
+        if (hasHDFSStorage) {
+          cleanExpiredAppDirsOnHDFS(appId)
+        }
         context.reply(ApplicationLostResponse(StatusCode.SUCCESS))
       }
     })
   }
 
+  private def cleanExpiredAppDirsOnHDFS(dir: String = ""): Unit = {
+    if (hadoopFs == null) {
+      hadoopFs = CelebornHadoopUtils.getHadoopFS(conf)
+    }
+    val hdfsWorkPath = new Path(conf.hdfsDir, conf.workerWorkingDir)
+    if (hadoopFs.exists(hdfsWorkPath)) {
+      if (!dir.isEmpty) {
+        // delete specific app dir on application lost
+        Utils.tryLogNonFatalError(hadoopFs.delete(new Path(hdfsWorkPath, dir), true))
+      } else {
+        val iter = hadoopFs.listStatusIterator(hdfsWorkPath)
+        while (iter.hasNext) {
+          val startTime = System.currentTimeMillis()
+          val fileStatus = iter.next()
+          if (!statusSystem.appHeartbeatTime.containsKey(fileStatus.getPath.getName)) {
+            Utils.tryLogNonFatalError(hadoopFs.delete(new Path(hdfsWorkPath, dir), true))
+            logInfo(
```

Comment:
this log message is not right then, I think we need a dedicated method here.

Reply:
Updated.